theia-docker/Dockerfile

```diff
@@ -4,24 +4,28 @@ ARG version=latest
 WORKDIR /home/theia
 ADD $version.package.json ./package.json
 ARG GITHUB_TOKEN
-RUN yarn && \
+RUN yarn --pure-lockfile && \
     yarn theia build && \
+    yarn --production && \
+    yarn autoclean --init && \
+    yarn autoclean --force && \
     rm -rf ./node_modules/electron && \
     yarn cache clean
 
 FROM node:8-alpine
 # See : https://github.com/theia-ide/theia-apps/issues/34
-RUN addgroup theia && \ 
+RUN addgroup theia && \
     adduser -G theia -s /bin/sh -D theia;
 RUN chmod g+rw /home && \
     mkdir -p /home/project && \
     chown -R theia:theia /home/theia && \
     chown -R theia:theia /home/project;    
 RUN apk add --no-cache git openssh bash
+ENV HOME /home/theia
 WORKDIR /home/theia
-COPY --from=0 /home/theia /home/theia
+COPY --from=0 --chown=theia:theia /home/theia /home/theia
 EXPOSE 3000
 ENV SHELL /bin/bash
 ENV USE_LOCAL_GIT true
 USER theia
-ENTRYPOINT [ "yarn", "theia", "start", "/home/project", "--hostname=0.0.0.0" ]
```

Comment:
You are right, `CMD` doesn't allow you to pass that arguments like that (should probably change `theia-full` to use `ENTRYPOINT` as well?). 

However, this would not matter for the case of this PR, since its using `--production`, it loses `theia` node bash script. So using either `CMD` or `ENTRYPOINT` or both, and adding additional arguments for the executable (in this case `node`) will not work since they are appended at the end of the command. The arguments passed are only usable by the `theia` and not `node`.

Reply:
But of course its not impossible, in order to pass arguments to `node`, you can simply use `NODE_OPTIONS` environment variable. Example:
```bash
docker run -it \
    -p 3000:3000 \
    --expose 9229 \
    -p 9229:9229 \
    -v "$(pwd):/home/project:cached" \
    -e NODE_OPTIONS='--inspect=0.0.0.0:9229' \
    theiaide/theia:next
```

Reply:
> ...The arguments passed are only usable by main.js and not node.

`theia start` only passes arguments to `main.js` and then to server workers, including `inspect` to handle debugging of forked processes in docker.

> But of course its not impossible, in order to pass arguments to node, you can simply use NODE_OPTIONS environment variable. 

The point is to avoid changing tweets, markdown files and so on which already explain how to use docker images. If it is only matter of the taste I would prefer to keep the old approach to avoid UX breaking.

Reply:
>  theia start only passes arguments to main.js and then to server workers, including inspect to handle debugging of forked processes in docker.

Yes in general that's right. But in details, correct me if im wrong, (after build) `theia` which is called by yarn is a node bash script which then passes the commands+arguments to `theia.ts` which then parses the command in this case `start`, then passes the arguments to `application-package-manager` which then parses the arguments and checks specifically for arguments that starts with `--inspect` (meaning no other nodejs options can be passes such as `--max-old-space-size`?) which then calls the `main.js`. So directly calling `main.js` bypasses all that.

One way this could be achieved if keeping this pr and without having to change anything on social aspects of theia(tweets) would be to create a bash script which parses node options and use that script as the entry point.

Reply:
A quick example of this using script on `ENTRYPOINT` to be able to keep the original approach:

File: `theia`(could be any name) as `ENTRYPOINT`
```
#!/bin/bash
for arg do
  shift
  if [[ $arg == --inspect* ]] ; then
   inspect="$arg"
  else
    set -- "$@" "$arg"
  fi
done
exec node $inspect /home/theia/src-gen/backend/main.js $@
```
The script above basically just checks for `--inspect` (`--inspect-brk` and `--inspect-port` works too) option.
Then on `Dockerfile` at the 2nd stage
```
...
COPY --from=0 --chown=theia:theia /home/theia /home/theia
COPY theia /usr/local/bin/theia
RUN chmod +x /usr/local/bin/theia
...
USER theia
ENTRYPOINT [ "theia", "/home/theia/src-gen/backend/main.js", "/home/project", "--hostname=0.0.0.0" ]
```
The script can be edited to accept other node options other than `--inspect`

Doing this we preserve the original approach and at the same time lessen the image size.

Reply:
I've tried with current state of PR, args are passed properly:
<img width="806" alt="screen shot 2018-10-29 at 09 09 41" src="https://user-images.githubusercontent.com/3082655/47636977-7b7b0700-db5a-11e8-8e80-f1adc2580e48.png">